src/ripple/consensus/LedgerTrie.h

```diff
@@ -0,0 +1,712 @@
+
+        Seq
+        end() const
+        {
+            return end_;
+        }
+
+        // Return the Span from (spot,end_]
+        Span
+        from(Seq spot)
+        {
+            return sub(spot, end_);
+        }
+
+        // Return the Span from (start_,spot]
+        Span
+        before(Seq spot)
+        {
+            return sub(start_, spot);
+        }
+
+        bool
+        empty() const
+        {
+            return start_ == end_;
+        }
+
+        //Return the ID of the ledger that starts this span
+        ID
+        startID() const
+        {
+            return ledger_[start_];
+        }
+
+        // Return the ledger sequence number of the first possible difference
+        // between this span and a given ledger.
+        Seq
+        diff(Ledger const& o) const
+        {
+            return clamp(mismatch(ledger_, o));
+        }
+
+        //  The Seq and ID of the end of the span
+        std::pair<Seq, ID>
+        tip() const
+        {
+            Seq tipSeq{end_ -Seq{1}};
+            return {tipSeq, ledger_[tipSeq]};
+        }
+
+    private:
+        Span(Seq start, Seq end, Ledger const& l)
+            : start_{start}, end_{end}, ledger_{l}
+        {
+            assert(start <= end);
+        }
+
+        Seq
+        clamp(Seq val) const
+        {
+            return std::min(std::max(start_, val), end_);
+        };
+
+        // Return a span of this over the half-open interval [from,to)
+        Span
+        sub(Seq from, Seq to)
+        {
+            return Span(clamp(from), clamp(to), ledger_);
+        }
+
+        friend std::ostream&
+        operator<<(std::ostream& o, Span const& s)
+        {
+            return o << s.tip().second << "(" << s.start_ << "," << s.end_
+                     << "]";
+        }
+
+        friend Span
+        merge(Span const& a, Span const& b)
+        {
+            // Return combined span, using ledger_ from higher sequence span
+            if (a.end_ < b.end_)
+                return Span(std::min(a.start_, b.start_), b.end_, b.ledger_);
+
+            return Span(std::min(a.start_, b.start_), a.end_, a.ledger_);
+        }
+    };
+
+    // A node in the trie
+    struct Node
+    {
+        Node() : span{}, tipSupport{0}, branchSupport{0}
+        {
+        }
+
+        Node(Ledger const& l) : span{l}, tipSupport{1}, branchSupport{1}
+        {
+        }
+
+        Node(Span s) : span{std::move(s)}
+        {
+        }
+
+        Span span;
+        std::uint32_t tipSupport = 0;
+        std::uint32_t branchSupport = 0;
+
+        std::vector<std::unique_ptr<Node>> children;
+        Node* parent = nullptr;
+
+        /** Remove the given node from this Node's children
+
+            @param child The address of the child node to remove
+            @note The child must be a member of the vector. The passed pointer
+                  will be dangling as a result of this call
+        */
+        void
+        erase(Node const* child)
```

Comment:
Have to think about the dangling pointer after this call. Maybe pass by `*&` and set the pointer to null after this call. Another option is to store `shared_pointers` and and `weak_ptrs` externally. Have to look at more code to decide.

Reply:
I'm fine nulling the pointer if you think it is worth it, but would prefer to avoid `shared_ptr`/`weak_ptr`. Since this is an internal detail class that is only used in one other spot, maybe I should just inline it there instead?

Reply:
If fine with this as-is. You're right, it's a private helper class. I'd love to think of a better name (`delete` would work if it wasn't reserved). At any rate, no change needed.